### DISCUSSION TITLE
SWR Optimization

### DISCUSSION THREAD

**Discussion by ghost** [vercel/swr#649] [General] [answered]
From what I understand, SWR essentially uses the fetcher function to poll your API at every interval of time in order to revalidate it's current state. However, how does that work with regards to memoization and optimization? Won't the refetch trigger a rerender?
Especially if you have dependent components, like TeamList > Team > TeamMember in which you have to prop drill IDs between levels you could have many needless rerenders.

**Comment by shuding** [comment, marked as answer]
> Won't the refetch trigger a rerender?

SWR uses 2 techniques to avoid unnecessary re-renders. In conclusion, the case you mentioned will not happen.

### Deep Comparison

Every time a "fetch" (`revalidate`) is triggered and when the data arrives, SWR uses [deep comparison](https://github.com/vercel/swr/blob/05755f83597643d8f386e1a33f1fa46325e04801/src/use-swr.ts#L444-L448) to check if the data has changed. If nothing changes, SWR will **not** update the state and no re-render will happen.

Currently, it uses [this lib](https://github.com/epoberezkin/fast-deep-equal) to check the data equality.

### Dependency Detection

A normal use case of SWR is that you have _some states_ being used by the component:

```js
const { data, isValidating } = useSWR(...)
```

The component will be rendered 3 times, caused by this 3 state changes:

`data=undefined, isValidating=false` → `data=undefined, isValidating=true` → `data="OK", isValidating=false`

However, if you are only using `data`, for example:

```js
const { data } = useSWR(...)
```

The component will be rendered 2 times:

`data=undefined` → `data="OK"`

because `isValidating` is never accessed. If you are interested in the implementation details: https://github.com/vercel/swr/pull/186.
